webserver/views/api.py

```diff
@@ -20,140 +19,30 @@
-        log_and_raise_bad_request("JSON document does not contain a valid metadata.track_name and/or track_metadata.artist_name", listen)
-
-    # Validate tags
-    if 'additional_info' in listen['track_metadata']:
-        tags = listen['track_metadata']['additional_info']['tags']
-        if len(tags) > MAX_TAGS_PER_LISTEN:
-            log_and_raise_bad_request("JSON document may not contain more than %d items in track_metadata.additional_info.tags" % MAX_TAGS_PER_LISTEN, listen)
-
-        for tag in tags:
-            if len(tag) > MAX_TAG_SIZE:
-                log_and_raise_bad_request("JSON document may not contain track_metadata.additional_info.tags longer than %d characters." % MAX_TAG_SIZE, listen)
 
 @api_bp.route("/listen/user/<user_id>", methods=["POST", "OPTIONS"])
 @crossdomain(headers="Authorization, Content-Type")
 def submit_listen(user_id):
-    """Endpoint for submitting a listen to ListenBrainz. Sanity check listen and then pass on to Kafka."""
+    """Endpoint for submitting a listen to ListenBrainz.
 
-    token = request.headers.get('Authorization')
-    if not token:
-        raise Unauthorized("You need to provide an Authorization header.")
-
-    try:
-        token = token.split(" ")[1]
-    except IndexError:
-        raise Unauthorized("Invalid Authorization header provided.")
-
-    user = db.user.get_by_mb_id(user_id)
-    if user is None:
-        raise Unauthorized("User %s is not known to MusicBrainz." % user_id)
-
-    if token != user['auth_token']:
-        raise Unauthorized("Invalid authorization token.")
+    Sanity check listen and then pass on to Kafka.
+    """
+    _validate_auth_header(user_id)
 
     raw_data = request.get_data()
     try:
         data = json.loads(raw_data.decode("utf-8"))
     except ValueError as e:
-        log_and_raise_bad_request("Cannot parse JSON document: %s" % e, raw_data)
-
-    # Sanity check the submission
-    try:
-        payload = data['payload']
-        if len(payload) == 0:
-            log_and_raise_bad_request("JSON document does not contain any listens.", payload)
-
-        if len(raw_data) > len(payload) * MAX_LISTEN_SIZE:
-            log_and_raise_bad_request("JSON document is too large. In aggregate, listens may not be larger than %d characters." % MAX_LISTEN_SIZE, payload)
+        _log_raise_400("Cannot parse JSON document: %s" % e, raw_data)
 
-        if data['listen_type'] not in ('playing_now', 'single', 'import'):
-            log_and_raise_bad_request("JSON document requires a valid listen_type key", payload)
-
-        if (data['listen_type'] == "single" or data['listen_type'] == 'playing_now') and len(payload) > 1:
-            log_and_raise_bad_request("JSON document contains more than listen for a single/playing_now. "
-                             "It should contain only one. ", payload)
-    except KeyError:
-        log_and_raise_bad_request("Invalid JSON document submitted.", raw_data)
+    payload = _get_listen_payload(data, raw_data)
```

Comment:
I can extract validation part into `validate_listen_payload` and put getting where it was. Would that work?